src/OpenLoco/src/Vehicles/Routing.cpp

```diff
@@ -683,19 +700,220 @@ namespace OpenLoco::Vehicles
+                        const auto cost = (baseCost * TrackData::getTrackCostFactor(tad.id())) / 256;
+                        totalCost += cost;
+                    }
+                }
+
+                if (flags & GameCommands::Flags::apply)
+                {
+                    bool invalidate = false;
+                    if (flags & GameCommands::Flags::ghost)
+                    {
+                        if (CompanyManager::getControllingId() == companyId)
+                        {
+                            elTrack->setHasGhostMods(true);
+                            invalidate = true;
+                        }
+                    }
+                    else
+                    {
+                        for (auto i = 0; i < 4; ++i)
+                        {
+                            if (trackModObjectIds & (1U << i))
+                            {
+                                elTrack->setMod(i, true);
+                            }
+                        }
+                        invalidate = true;
+                    }
+                    if (invalidate)
+                    {
+                        Ui::ViewportManager::invalidate(trackLoc, elTrack->baseHeight(), elTrack->clearHeight(), ZoomLevel::half);
+                    }
+                }
+            }
+        }
+
+        if (modSelection == 1)
+        {
+            return interest.trackAndDirection & Track::AdditionalTaDFlags::hasSignal;
+        }
+        return false;
+    }
+
+    ApplyTrackModsResult applyTrackModsToTrackNetwork(const World::Pos3& pos, Vehicles::TrackAndDirection::_TrackAndDirection trackAndDirection, CompanyId company, uint8_t trackType, uint8_t flags, uint8_t modSelection, uint8_t trackModObjIds)
+    {
+        ApplyTrackModsResult result{};
+        result.cost = 0;
+        result.allPlacementsFailed = true;
+        result.networkTooComplex = false;
+
+        if (modSelection == 0)
+        {
+            LocationOfInterest interest{ pos, trackAndDirection._data, company, trackType };
+            applyTrackModToTrack(interest, flags, nullptr, modSelection, trackType, trackModObjIds, result.cost, company, result.allPlacementsFailed);
+            return result;
+        }
+
+        LocationOfInterestHashMap interestHashMap{ kTrackModHashMapSize };
+
+        auto filterFunction = [flags, modSelection, trackType, trackModObjIds, &result, company, &interestHashMap](const LocationOfInterest& interest) {
+            return applyTrackModToTrack(interest, flags, &interestHashMap, modSelection, trackType, trackModObjIds, result.cost, company, result.allPlacementsFailed);
+        };
+        findAllTracksFilterTransform(interestHashMap, TrackNetworkSearchFlags::unk0, pos, trackAndDirection, company, trackType, filterFunction, kNullTransformFunction);
+        result.networkTooComplex = interestHashMap.count >= interestHashMap.maxEntries;
```

Comment:
I'm pretty sure this will never be hit as its off by one. This is how vanilla is. Would be interesting to test (does anyone have an 4096+ element track network that can have track mods applied to it?

Reply:
Sounds like a challenge for one of the players on Discord that build these incredibly complicated networks.

Reply:
I may have a map that complex lying around, how can I easily count the number of track pieces in the game?

Reply:
Its easy enough in code. Assuming all tracks interlink then you just iterate the whole map counting for track elements with sequence 0. Or you just run this code and see what number it gets up to.

Reply:
Was this supposed to be 4096?
![image](https://github.com/OpenLoco/OpenLoco/assets/7483209/07ccfeff-937c-446d-a91b-693cfa8f4b24)

Reply:
This is what I get though with a track network of about 7k pieces:

![image](https://github.com/OpenLoco/OpenLoco/assets/7483209/7c7a25a3-f300-4c3c-ada7-9bfb0be34b57)

![image](https://github.com/OpenLoco/OpenLoco/assets/7483209/7c81fee0-4009-41e0-8e71-33a790a8478d)

No crash, which is good.

Reply:
So it looks like interestHashMap cannot exceed 3996 entries, so that check on line 916 can just be `==`, or alternatively allow the map to be the full size

Reply:
Ah ok makes sense, 10 entries is nothing to care about here.

The error message is a little grammatically off though, I'd write `Warning: Too much track! Some selected track has not been upgraded`. Looks like it's an existing string though, but might be nice to fix it up in this PR since it's used here?